Add `tfp.math.betaincinv`

leandrolcampos opened the pull request:
This PR provides an implementation for `tfp.math.betaincinv`, the inverse of the regularized incomplete beta function (`tfp.math.betainc`) with respect to `x`. It addresses the issue #1581.

This implementation is based on ideas and equations available in the following references:
[1] Milton Abramowitz and Irene A. Stegun. Handbook of Mathematical Functions with Formulas, Graphs, and Mathematical Tables. US Government Printing Office, 1964 (reprinted 1972). [[Link](https://archive.org/details/AandS-mono600)]
[2] William Press, Saul Teukolsky, William Vetterling and Brian Flannery. Numerical Recipes: The Art of Scientific Computing. Cambridge University Press, 2007 (Third Edition). [[Link](http://numerical.recipes/book/book.html)]
[3] John Maddock, Paul A. Bristow, et al. The Incomplete Beta Function Inverses. [[Link](https://www.boost.org/doc/libs/1_79_0/libs/math/doc/html/special.html)]
[4] Stephen L. Moshier. Cephes Mathematical Library. [[Link](https://netlib.org/cephes/)]

The expression of these ideas and equations as source code is my own.

In the following notebook, I measure and present the numerical accuracy of this implementation:
- [My implementation of `betaincinv`](https://gist.github.com/leandrolcampos/767818366dd8cbdffb6189aa00847d34)

As references (especially for single-precision results), I also measure and present the numerical accuracy of the following methods:
- [Cephes implementation of `betaincinv` for `float32` [my port]](https://gist.github.com/leandrolcampos/5dfbc66753310598ef531f59901a343b)
- [TFP implementation of `igammainv`](https://gist.github.com/leandrolcampos/b19177cac0e4f3f0b962edca265ec8ab)

In addition to implementing the function `tfp.math.betaincinv`, this PR also fixes a bug in the implementation of the partial derivative of `tfp.math.betainc` with respect to `x`: it's not true that this partial is equal to `0` whenever the parameter `x` is equal to `0` or `1`.

leandrolcampos commented:
I'm worried about dtype support in the implementations for `tfp.math._betainc_partials` and `tfp.math.betaincinv`. They only support `float32` and `float64` by design, like the TF C++ and XLA implementations for `betainc`. But there is an important difference between these last two implementations: the former throws an error when the input dtype is `float16` or `bfloat16`, while the latter converts a `float16` or `bfloat16` input to `float32` and then, after the computations, converts the result back to the original dtype [[link](https://github.com/tensorflow/tensorflow/blob/af2b6266a53b8bfa8535c84860313b11651b363a/tensorflow/compiler/xla/client/lib/math.cc#L1737-L1862)].

Thinking about consistency, should we cast a `float16` or `bfloat16` input to `float32` and then, after the computations, convert the result back to the original dtype?

leandrolcampos commented:
Hi, @srvasude!

I hope this message finds you very well.

When you have time, could you review this PR, please?

All the best,

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -521,6 +545,332 @@ def betainc(a, b, x, name=None):
+  inv_a = tf.math.reciprocal(a)
+  inv_b = tf.math.reciprocal(b)
+  result_for_small_a_or_b = tf.where(
+      tf.math.less_equal(y, (integral_approx_part_a / integral_approx)),
```
y <= integral_approx_part_a / integral_approx. Can use `<=` and `>=` overrides here and elsewhere

leandrolcampos replied:
Done (here and elsewhere).

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -521,6 +545,332 @@ def betainc(a, b, x, name=None):
+      tf.math.less_equal(y, (integral_approx_part_a / integral_approx)),
+      tf.math.exp(tf.math.xlogy(inv_a, y) + tf.math.xlogy(inv_a, a) +
+          tf.math.xlogy(inv_a, integral_approx)),
+      one - tf.math.exp(tf.math.xlog1py(inv_b, -y) + tf.math.xlogy(inv_b, b) +
```
Can use `-tf.math.expm1` instead of `one - exp(...)`

leandrolcampos replied:
Done.

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -521,6 +545,332 @@ def betainc(a, b, x, name=None):
+  numpy_dtype = dtype_util.as_numpy_dtype(dtype)
+  tiny = tf.constant(np.finfo(numpy_dtype).tiny, dtype=dtype)
+  eps = tf.constant(np.finfo(numpy_dtype).eps, dtype=dtype)
+  one = tf.constant(1., dtype=dtype)
```
FWIW can use `numpy_dtype` instead (just my own style preference, feel free to keep as is):

`one = tf.constant(1., dtype=dtype)` vs
`one = numpy_dtype(1.)`

leandrolcampos replied:
Done (here and elsewhere). Thanks for sharing this!

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -521,6 +545,332 @@ def betainc(a, b, x, name=None):
+  three = tf.constant(3., dtype=dtype)
+  five = tf.constant(5., dtype=dtype)
+  six = tf.constant(6., dtype=dtype)
+  min_log = tf.math.log(
```
`min_log = numpy_dtype(np.finfo(numpy_dtype).minexp * np.log(2.))`
`max_log = numpy_dtype((np.info(numpy_dtype).maxexp - 1.) * np.log(2.))`

leandrolcampos replied:
Done.

srvasude commented on tensorflow_probability/python/math/special_test.py:
```diff
@@ -194,6 +194,37 @@ def testBetaincBroadcast(self):
+  @test_util.disable_test_for_backend(
+      disable_numpy=True, disable_jax=False,
+      reason="Numpy does not support brain floating point.")
+  def testBetaincBrain(self):
```
`testBetaincBFloat16`, and probably change the reason to `bfloat16`

leandrolcampos replied:
Done (here and elsewhere).

srvasude commented on tensorflow_probability/python/math/special.py:
```diff
@@ -521,6 +545,332 @@ def betainc(a, b, x, name=None):
+  #   xg = (a * y * Beta(a, b)) ** (1 / a) .
+  log_xg = tf.math.xlogy(inv_a, a) + tf.math.xlogy(inv_a, y) + (
+      inv_a * lbeta(a, b))
+  xg = tf.math.exp(tf.clip_by_value(log_xg, min_log, max_log))
```
There seems to be quite a bit of clipping in this code (here and a few places below). Are these necessary / can they be removed in favor of `tf.where` guarding code earlier up?

leandrolcampos replied:
There were four uses of `tf.clip_by_value` in the code (I replaced two of them). They handled cases that depend on the combination of values for the parameters `a`, `b`, and `y`. I don't know if it's possible (and convenient) to list and describe all these problematic combinations of values and then use `tf.where` to guard the code from the respective cases.

The first use of `tf.clip_by_value` (line 629) avoided overflow in `tf.math.exp(log_xg)`: for example, when `a` and `b` are very small and `x` is close enough to `1`. But in this case, I replaced `tf.clip_by_value` with `tf.math.minimum`.

The second use of `tf.clip_by_value` (line 641) ensures that the result of `_betaincinv_initial_approx` is always greater than `0` and less than `1` and, consequently, that the condition `initial_low < initial_candidate < initial_high` is always satisfied. For example, if

```
a, b, y, dtype = (3.5565727, 0.011408293, 0.130523313, float32)
```

then `initial_candidate == 1.` and `initial_candidate == initial_high`.

The reason for the third use of `tf.clip_by_value` (line 714) is commented out in the code.

The last use of `tf.clip_by_value` (line 764) prevented `one - result` from being rounded up to `1` when `use_symmetry_relation` and `result < eps`. But in this case, I replaced `tf.clip_by_value` with `tf.math.maximum`.

srvasude replied:
Thanks for the explanation, those uses seem pretty reasonable to me.

srvasude commented on tensorflow_probability/python/math/special_test.py:
```diff
@@ -263,15 +294,16 @@ def betainc_partials(a, b, x):
 
     for partial in self.evaluate(betainc_partials(a, b, x)):
-      self.assertAllClose(np.full_like(a, np.nan), partial)
+      self.assertAllEqual(np.full_like(a, np.nan), partial)
```
can use `self.assertAllNan`

leandrolcampos replied:
Done (here and elsewhere).

srvasude reviewed (commented):
Sorry for the delay, and thanks for this contribution! Overall looks great to me, with a few style comments and one comment (about clipping).

leandrolcampos commented:
Hi @srvasude,

Thank you very much for your time reviewing this PR. It's always a pleasure to learn from you!

Please let me know if I've properly addressed all your comments.

All the best,
